### PR TITLE
Extend preprocessor logic

### PR DESCRIPTION
**Pull request by zaucker**
- add ?or and ?and to ?if clause
- add tests
- add tests for unsupported usage

### CODE REVIEW

**Review by jnthn — COMMENTED** [review]
This isn't a terribly elegant solution; there's:

* Reparsing (the split, the re-identitying the negation)
* Code duplication between the cases
* A bit too much strictness on whitespace around the operators (hard-coded spaces instead of `\h+`)
* Potential bugs because `.+` in Raku includes newlines
* Fairly arbitrary of exclusion of `elsif` support for the feature

In my opinion the `?or` is also overkill, since the `?` already said we're in a preproc directive, so I think it can be just `or` there.

I think it would overall be much cleaner to write one rule for all the and/or/single cases like this:

```
when /^ '?if' $<parts>=(\h+ [$<negate>='!']? <option=.ident>)+ % [\h+ $<op>=< and or >] \h* $/ {
}
```

That gets you a neat, structured, match tree. While we could go and do proper precedence, let's for now accept only `or`s or only `and`s; we can detect the mixing with:

```
unless ‌‌all([eq] $<op>) {
    die X::Agrammon::Preprocessor::MixedAndOr.new: :line($number + 1);
}
```

From there it depends if a short-circuit is desired or not. Let's assume not. Then we can do:

```
my @enabled-parts = $<parts>.map: -> $/ {
     $<negate> ?? .not !! .so given %options{~$<option>}
}
```

And then collapse them to a single boolean value:

```
my $enabled = $<op> && $<op>[0] eq 'or'
    ?? ?any(@enabled-parts)
    !! ?all(@enabled-parts);
```

Noting that the `all` handles the case of no operator, because `?all($bool)` is the same as `$bool`.

Given how much less additions this causes, it's probably now not so terrible to implement it for `elsif` too (especially as the processing can be extracted to a `sub`).

**Comment by zaucker** [comment]
> This isn't a terribly elegant solution; there's:

Re-implemented it with your suggestions, thanks!

File: lib/Agrammon/Preprocessor.pm6 [hunk 1]
```diff
@@ -64,18 +80,18 @@ sub preprocess(Str $source, %options --> Str) is export {
...
                     @open.push(OpenDirective.new(:$start-line, :$enabled, :$matched, :accept-elsif));
                 }
-                when /^ '?elsif' \h+ [$<negate>='!']? <option=.ident> \h* $/ {
+                when /^ '?elsif' $<parts>=(\h+ [$<negate>='!']? <option=.ident>)+ % [\h+ $<op>=< and or >] \h* $/ {
                     if @open.pop -> $prev-part {
                         unless $prev-part.accept-elsif {
                             die X::Agrammon::Preprocessor::ElsifAfterElse.new: :line($number + 1);
                         }
-                        my $enabled = so !$prev-part.matched && ($<negate> ?? .not !! .so given %options{~$<option>});
+                        my $enabled = so !$prev-part.matched && process-match($<parts>, $<negate>, %options, $<op>, $<option>, $number);
```
Comment (jnthn) [inline lib/Agrammon/Preprocessor.pm6:94]: It's a nitpick but the `so` is not needed here given `!` and `process-match` both return booleans.

**Review by jnthn — COMMENTED** [review]
About fine now, one bit of dead code to remove.

File: lib/Agrammon/Preprocessor.pm6 [hunk 2]
```diff
@@ -34,19 +34,35 @@ class X::Agrammon::Preprocessor::ElsifAfterElse does X::Agrammon::Preprocessor {
...
+        "Invalid preprocessor ?and and ?or in same ?if found at line $!line"
+    }
+}
+
+#| Exception thrown when ?elsif with multiple options
+class X::Agrammon::Preprocessor::ElsifMultiOptions does X::Agrammon::Preprocessor {
+    method message() {
+        "Invalid preprocessor ?elsif with multiple options found at line $!line"
+    }
+}
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/Preprocessor.pm6:49]: Removed.

File: lib/Agrammon/Preprocessor.pm6 (hunk 1 above)
Comment (zaucker) [reply lib/Agrammon/Preprocessor.pm6:94]: Removed.

**Comment by zaucker** [comment]
One question: why is the `|$op` needed here
```
sub process-match($parts, $negate, %options, $op, $option, $number --> Bool) {
    unless all([eq] |$op) {
        die X::Agrammon::Preprocessor::MixedAndOr.new: :line($number + 1);
    }
```
while it is not needed if the test ist done inside `preprocess()` like this
```
    unless all([eq] |$<op>) {
```
And how can I tell the difference between `$<op>` in the caller and `$op` in the called sub?

**Comment by jnthn** [comment]
`$<op>` is syntax sugar for `$/<op>`, and what you get back is a `List`, which `[eq]` then iterates.

Since `$` means item, then when receiving it into `$op` it gets wrapped in a `Scalar` container, and then `[eq]` sees it as a single item.

You might prefer to use `@op` in the signature instead, since `$<op>` gives a `List` of matches back, so that's the better sigil for it.

**Review by jnthn — APPROVED** [review]

**Comment by zaucker** [comment]
Thanks for the explanation. Now I understand the `$[]` versus `[]` which `dd` showed (wrapping in Scalar container).

I guess I should use `@` for the other lists as well.

Will update the PR.
> $<op> is syntax sugar for $/<op>, and what you get back is a List, which [eq] then iterates.
> 
> Since $ means item, then when receiving it into $op it gets wrapped in a Scalar container, and then [eq] sees it as a single item.
> 
> You might prefer to use @op in the signature instead, since $<op> gives a List of matches back, so that's the better sigil for it.

**Comment by zaucker** [comment]
I realized that I hadn't quite understood the nesting of the regex match and thus passed useless (`Nil`) parameters to `process-match()`. Removed them and used `@` for the lists being passed.

**Review by jnthn — APPROVED** [review]

**Pull request merged**